doc/man/man1/gutil.1
.\" $Id: gutil.1,v 1.1 2002/02/28 19:34:24 engelhar Exp $
.TH gutil 1 "26 February 2002" "Web100 Userland" "Web100"
.SH NAME
gutil \- a \fIGTK+\fR based graphical front-end to the Web100 connection
statistics.
.SH SYNOPSIS
.B gutil
.I widget_name
[\fIconnection_id\fR]
.SH DESCRIPTION
\fBgutil\fR is a graphical front end which allows users to read, monitor,
and manipulate the values of Web100 variables on connections.
\fBgutil\fR internally contains many different tools to perform these
actions, and the parameter \fIwidget_name\fR specifies which tool you
want to use.  You can optionally provide a connection identifier,
\fIconnection_id\fR to specify which connection you wish to manipulate.
.PP
The table below shows the allowable values of \fIwidget_name\fR followed
by the widget's purpose.
.TP
\fBdtb\fR
a universal tool box that allows you to launch the other tools
.TP
\fBavd_list\fR
a continuously-updating list view of the variables in a single connection
.TP
\fBavd_table\fR
a continuously-updating table view of the variables in a single connection
.TP
\fBcpr\fR
monitors a few key parameters of a connection
.TP
\fBrtuner\fR
allows you to manipulate the \fIRTuneMode\fR variable of a connection
.TP
\fBstuner\fR
allows you to manipulate the \fISTuneMode\fR variable of a connection
.TP
\fBtriage\fR
allows you to watch \fISndLimTimeSender\fR, \fISndLimTimeCwnd\fR, and
\fISndLimTimeRwin\fR
.SH SEE ALSO
.BR readvar (1),
.BR readall (1),
.BR deltavar (1),
.BR writevar (1),
.BR web100 (7)


NEWS
News
====

    This document gives an overview of the major changes of the software since
the last release.

  Alpha 1.2pre2:
    o Fixed the binary interface incompatibilities that were introduced
      in Alpha 1.2pre1.  This means that programs which were compiled
      against Alpha 1.1 should continue to work unchanged with the
      Alpha 1.2pre2 libweb100.

  Alpha 1.2pre1:
    o NOTE: The binary interface to libweb100 has changed slightly.  This
      means that existing applications must be recompiled and relinked
      with libweb100.  However, it is very likely that the recompile will
      not require changes to existing source code.
    o Added support for both IPv4 and IPv6 types from the kernel (previous
      versions of the library would crash if the kernel returned IPv6
      values).
    o Added support for logging values from the kernel
    o Addition of 'readvarm' and 'readconn' scripts
    o Addition of man pages for many functions, such as the snapshot and
      logging functions (some are only stubs).
    o Many GUI improvements and cleanups
    o Updates and cleanups to the developer's and user's guide.
    o Removed some items from web100.h that were not meant to be public
    o Support building on systems which do not have socklen_t

  Alpha 1.1:
    o GUI fixes: Renamed "util" to "gutil", fixed the vdt display,
      consolidation of code, miscellaneous bug and display fixes.
    o writevar fixed so that it works for integers
    o Added web100_connection_from_socket() and web100_value_to_textn()
      to libweb100.
    o Bug fixes in libweb100, such as for _find() and _lookup()
      functions.
    o Documentation updates and additions, such as the developer's
      guide, user's guide, and man pages for some libweb100 API
      functions (see "man libweb100").

  Alpha 1.0:
    o First public release


util/gui/gutil.c
#include <stdlib.h>
#include <unistd.h>
#include <stdio.h>
#include <ctype.h>
#include <gtk/gtk.h>
#include "web100.h"
#include "web100sync.h"
#include "web100obj.h"
#include "sockset_wgt.h"
#include "avd_list_wgt.h"
//#include "avd_table_wgt.h"
#include "cpr_wgt.h"
#include "dtb_wgt.h"
#include "rtuner_wgt.h"
#include "stuner_wgt.h"
#include "triage_wgt.h"
#include "vdt_wgt.h"

#if 0	/* For debugging unaligned traps on alpha */
#include <asm/sysinfo.h>
#include <asm/unistd.h>

static int setsysinfo(unsigned long op, void *buffer, unsigned long size,
		      int *start, void *arg, unsigned long flag)
{
	syscall(__NR_osf_setsysinfo, op, buffer, size, start, arg, flag);
}

void trap_unaligned(void)
{
	unsigned int buf[2];
	buf[0] = SSIN_UACPROC;
	buf[1] = UAC_SIGBUS | UAC_NOPRINT;
	setsysinfo(SSI_NVPAIRS, buf, 1, 0, 0, 0);
}
#endif

Web100Sync *web100sync;
Web100Obj *web100obj;

void launch_util (char *name, Web100Obj *web100obj, char varname[WEB100_VARNAME_LEN_MAX], gboolean master)
{
static  GtkWidget *window, *vbox, *ur_sockset, *avd_list, *avd, *cpr, *dtb, *rtuner, *stuner, *triage, *vdt;

  window = gtk_window_new (GTK_WINDOW_TOPLEVEL);
  gtk_window_set_policy (GTK_WINDOW(window), TRUE, TRUE, FALSE); 
  gtk_window_set_title (GTK_WINDOW (window), name);

  if (master)
    gtk_signal_connect (GTK_OBJECT (window), "destroy",
                        GTK_SIGNAL_FUNC (gtk_main_quit), NULL);
  else // if slave
    gtk_signal_connect (GTK_OBJECT (window), "destroy",
                        GTK_SIGNAL_FUNC (gtk_widget_destroy), NULL); 
  

  gtk_container_set_border_width (GTK_CONTAINER (window), 10);

  vbox = gtk_vbox_new (FALSE, 0); 
  gtk_container_add (GTK_CONTAINER (window), vbox);
  gtk_widget_show (vbox);

  ur_sockset = sockset_new (web100obj, TRUE);
  gtk_box_pack_start (GTK_BOX(vbox), ur_sockset, FALSE, FALSE, 0);
  gtk_widget_show (ur_sockset);

  if (!strncmp ("avd_list", name, 8)) {
    avd_list = avd_list_new (web100obj);
    gtk_box_pack_start (GTK_BOX(vbox), avd_list, TRUE, TRUE, 0);
    gtk_widget_show (avd_list);

    gtk_signal_connect (GTK_OBJECT (ur_sockset),
       	"web100obj_changed",
       	GTK_SIGNAL_FUNC (avd_list_sockset_listen),
       	(gpointer) avd_list); 
  } 
#if 0
  if (!strncmp ("avd_table", name, 9)) {
    avd = avd_table_new (web100obj);
    gtk_box_pack_start (GTK_BOX(vbox), avd, TRUE, TRUE, 0);
    gtk_widget_show (avd);
  
    gtk_signal_connect (GTK_OBJECT (ur_sockset),
                        "web100obj_changed",
                        GTK_SIGNAL_FUNC (avd_table_sockset_listen),
                        avd); 
  } 
#endif
  if (!strncmp ("cpr", name, 3)) {
    cpr = cpr_new (web100obj);
    gtk_box_pack_start (GTK_BOX(vbox), cpr, TRUE, TRUE, 0);
    gtk_widget_show (cpr);

    gtk_signal_connect (GTK_OBJECT (ur_sockset),
                        "web100obj_changed",
                        GTK_SIGNAL_FUNC (cpr_sockset_listen),
                        cpr); 
  } 
  if (!strncmp ("dtb", name, 3)) {
    dtb = dtb_new (web100obj);
    gtk_box_pack_start (GTK_BOX(vbox), dtb, TRUE, TRUE, 0);
    gtk_widget_show (dtb);

    gtk_signal_connect (GTK_OBJECT (ur_sockset),
                        "web100obj_changed",
                        GTK_SIGNAL_FUNC (dtb_sockset_listen),
                        dtb); 
  } 
  if (!strncmp ("rtuner", name, 6)) {
    rtuner = rtuner_new (web100obj);
    gtk_box_pack_start (GTK_BOX(vbox), rtuner, TRUE, TRUE, 0);
    gtk_widget_show (rtuner);

    gtk_signal_connect (GTK_OBJECT (ur_sockset),
                        "web100obj_changed",
                        GTK_SIGNAL_FUNC (rtuner_sockset_listen),
                        rtuner); 
  }
  if (!strncmp ("stuner", name, 6)) {
    stuner = stuner_new (web100obj);
    gtk_box_pack_start (GTK_BOX(vbox), stuner, TRUE, TRUE, 0);
    gtk_widget_show (stuner);

    gtk_signal_connect (GTK_OBJECT (ur_sockset),
                        "web100obj_changed",
                        GTK_SIGNAL_FUNC (stuner_sockset_listen),
                        stuner); 
  } 
  if (!strncmp ("triage", name, 6)) {
    triage = triage_new (web100obj);
    gtk_box_pack_start (GTK_BOX(vbox), triage, TRUE, TRUE, 0);
    gtk_widget_show (triage);

    gtk_signal_connect (GTK_OBJECT (ur_sockset),
                        "web100obj_changed",
                        GTK_SIGNAL_FUNC (triage_sockset_listen),
                        triage); 
  }
  if (!strncmp ("vdt", name, 3)) {
    vdt = vdt_new (web100obj, varname);
    gtk_box_pack_start (GTK_BOX(vbox), vdt, TRUE, TRUE, 0);
    gtk_widget_show (vdt);

    gtk_signal_connect (GTK_OBJECT (ur_sockset),
       	"web100obj_changed",
       	GTK_SIGNAL_FUNC (vdt_sockset_listen),
       	vdt); 
  }
  gtk_widget_show (window);
}

int main(int argc, char *argv[] )
{ 
  extern int optind;
  int option;
  
  gtk_init(&argc, &argv);

  gtk_rc_parse(WEB100_CONF_DIR "/web100.rc");
  while ((option = getopt(argc, argv, "lr")) != -1)
    switch(option)
    {
      case 'l':
       	gtk_rc_parse(WEB100_CONF_DIR "/lcl.rc"); 
	break;
      case 'r':
       	gtk_rc_parse(WEB100_CONF_DIR "/rmt.rc"); 
	break;
      default: 
       	break;
    } 

  if ((web100sync = (Web100Sync *) web100sync_new (NULL)) == NULL) { 
    exit(1);
  }

  if(argv[optind]) {
    if (argv[optind+1]) { 
      if (argv[optind+2]) {
       	web100obj = WEB100_OBJ (web100obj_new(web100sync, argv[optind+2]));
       	launch_util(argv[optind], web100obj, argv[optind+1], TRUE);
      }
      else {
       	web100obj = WEB100_OBJ (web100obj_new(web100sync, argv[optind+1]));
       	launch_util(argv[optind], web100obj, NULL, TRUE); 
      }
    }
    else {
      if (isdigit(argv[optind][0])) {
	web100obj = WEB100_OBJ (web100obj_new(web100sync, argv[optind]));
       	launch_util("dtb", web100obj, NULL, TRUE);
      }
      else {
	web100obj = WEB100_OBJ (web100obj_new(web100sync, NULL));
       	launch_util(argv[optind], web100obj, NULL, TRUE);
      }
    }
  }
  else {
    web100obj = WEB100_OBJ (web100obj_new(web100sync, NULL));
    launch_util("dtb", web100obj, NULL, TRUE);
  }

  gtk_main ();

  return 0;
} 
